### PR TITLE
platform(general): Apply licensing from platform

### PR DESCRIPTION
**Pull request by mikeurbanski1**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

Adds logic to get the licensing details from the platform and run the correct runners.

Resource licensing: run all runners, as today.
Dev licensing: only run runners mapped to subscribed modules.

Secrets logic (i.e. whether to run all secrets checks, etc) will also come from the platform, but is handled separately. This change only affects whether the secrets runner runs at all.

If the user specifies one framework using `--framework`, and that framework is not enabled, they get a clear error message (as opposed to blank results). If they specify multiple frameworks and all are disabled, they get a similar message. If they specify multiple frameworks and only some are disabled, they get messages for the disabled ones.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my feature, policy, or fix is effective and works
- [X] New and existing tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/common/integration_features/test_licensing_integration.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        self.assertFalse(licensing_integration.open_source_only)
+
+        instance.customer_run_config_response = {
+            'platformLicense': {
+                'modules': {m: True for m in module_keys},
+                'billingPlan': 'RESOURCE_BASED'
+            }
+        }
+        licensing_integration.pre_scan()
+        self.assertFalse(licensing_integration.open_source_only)
```
Comment (nimrodkor) [inline tests/common/integration_features/test_licensing_integration.py:163]: @mikeurbanski1 These should be separate tests IMO

File: checkov/common/bridgecrew/integration_features/features/licensing_integration.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            logging.debug('Found customer run config and using it for licensing')
+            license_details = self.bc_integration.customer_run_config_response.get(LICENSE_KEY)
+            logging.debug(f'User license details: {license_details}')
+
+            self.open_source_only = False
+            # the API will return True for all modules if they are on resource mode, so we don't actually need the billing plan explicitly here
+            self.enabled_modules = [CustomerSubscription(m) for m, e in license_details.get(MODULES_KEY).items() if e]
+            self.billing_plan = BillingPlan(license_details[BILLING_PLAN_KEY])
+
+    def is_runner_valid(self, runner: str) -> bool:
```
Comment (nimrodkor) [inline checkov/common/bridgecrew/integration_features/features/licensing_integration.py:51]: @mikeurbanski1 Just making sure - we are leveraging here the runner's `check_type`, right?
Can we make that explicit, i.e.:
```suggestion
    def is_runner_valid(self, runner_check_type: str) -> bool:
```

(Need to change in other places too...)

**Review by nimrodkor — COMMENTED** [review]
Very nice @mikeurbanski1 !

File: checkov/common/bridgecrew/licensing.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+from checkov.common.bridgecrew.code_categories import CodeCategoryType
+
+
+class CustomerSubscription(Enum):
+    IAC = "IAC"
+    SCA = "SCA"
+    SECRETS = "SECRETS"
+
+
+class BillingPlan(Enum):
```
Comment (gruebel) [inline checkov/common/bridgecrew/licensing.py:12]:
```suggestion
class CustomerSubscription(str, Enum):
    IAC = "IAC"
    SCA = "SCA"
    SECRETS = "SECRETS"


class BillingPlan(str, Enum):
```

File: checkov/common/bridgecrew/licensing.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    DEVELOPER_BASED = "DEVELOPER_BASED"
+    RESOURCE_BASED = "RESOURCE_BASED"
+
+
+SubscriptionCategoryMapping = {
+    CustomerSubscription.IAC: [CodeCategoryType.IAC, CodeCategoryType.SUPPLY_CHAIN],
+    CustomerSubscription.SCA: [CodeCategoryType.OPEN_SOURCE, CodeCategoryType.IMAGES],
+    CustomerSubscription.SECRETS: [CodeCategoryType.SECRETS]
+}
```
Comment (gruebel) [inline checkov/common/bridgecrew/licensing.py:21]:
```suggestion
SubscriptionCategoryMapping = {
    CustomerSubscription.IAC: (CodeCategoryType.IAC, CodeCategoryType.SUPPLY_CHAIN),
    CustomerSubscription.SCA: (CodeCategoryType.OPEN_SOURCE, CodeCategoryType.IMAGES),
    CustomerSubscription.SECRETS: (CodeCategoryType.SECRETS,)
}
```
if the list are static, then a tuple is more performant

**Review by gruebel — COMMENTED** [review]
great job 🍻

**Comment by mikeurbanski1** [comment]
Converting to draft so we don't merge it before the platform changes

**Comment by mikeurbanski1** [comment]
Because we are trying to get better about error messages for users, I also made some updates to explicitly fail if all the runners are invalid, and to log at warning (not info, which they won't see) if some runners are invalid, but only if they used `--framework` (we will still log skipped runners at info level)

**Review by reuvenz1993 — APPROVED** [review]

**Comment by reuvenz1993** [comment]
LGTM great job!

**Review by gruebel — APPROVED** [review]
🎄

**Pull request merged**